### ISSUE TITLE
Add particle limit setting

### ISSUE THREAD

**Issue by epicEaston197** [Yellow-Dog-Man/Resonite-Issues#1764]
### Is your feature request related to a problem? Please describe.

Sometimes people can get a little too crazy in the world and I end up freezing because they used way too many particles

### Describe the solution you'd like

I would like to have a setting to limit how many particles can exist at once

### Describe alternatives you've considered

none

### Requesters

@epicEaston197

**Comment by Frooxius** [comment]
Using Unity system doesn't matter here, but this is a bit of a trickier issue. 

The limit that is okay for one particle system, might not be enough for another and completely break the system/content. 

E.g. if you have particle system tuned to do a constant stream of particles and you suddenly lower the max limit, now it'll start acting in "bursts" with gaps, which is not desirable - we'd need to scale the emission rate a well appropriately. 

But even that could possibly break some effects. We'll need to think about how to achieve something like this with minimal breakage.

**Comment by epicEaston197** [comment]
No there are situations to where people are messing around with particle systems and they accidentally add an extra 0 resulting in the game just crashing or freezing out of nowhere which is not preferable I want a hard limit on particle systems because someone using a million particles is just outrageous

**Comment by ultrawidegamer** [comment]
I agree with @epicEaston197 We need a way to set a hard user limit on particles, even if it's restrictive. Users should be able to control the maximum number of particles their computer can handle before it stops creating more particles in the world. Alternatively, they should be able to set it to zero to disable particles entirely on a low-performance computer. A possible workaround could be similar to the "Skin Weights" rendering setting. In the additional information section, users could be warned that reducing the value too much might unintentionally degrade the viewing experience.

**Comment by epicEaston197** [comment]
This seriously needs to be prioritized as it crashed someones entire computer

here's a quote from the user from the modding discord

> This is the second time I had to suffer from this. Basically I was trying to hang out and chill with others and suddenly a lot of particles explode and made my entire computer crash hard (full black screen). I wish for a mod that can limit the number of particles at a given time (even if that means I wont load, see, hear them locally from my client).
> 
> Since I crashed, I have no way of knowing what item has caused it, I can only recall the last few moments of my experience. I remember seeing a lot of flame particles and loud sounds stacked together in a sudden burst (similar to the component "ViolentAprilFoolsExplosion"). Then my headset cut out and then all my monitors blacked out. I hear a beep from my computer and it restarts again.
> 
> My best guess could be from the item "<color=red>=== Yeeted ===</color>" in the logs, but I dont know, I crashed before I could do anything about it. Also sorry for the huge log file, it was a long play session and I might as well use the latest log file (I was unprepared for something like this again, so no clean logs).

**Comment by NukiRaccoon** [comment]
I know this is old, but I wanted to bring this back up on the way to making a related request...  Perhaps this should have a standard cloud variable?  Like, some "high / med / low" type thing that people can collaboratively utilize?  I would definitely use and encourage using this variable to limit particle systems on people who indicated their computers couldn't handle that many...

...then again, is this still an issue after photon dust?  If the particle system is so much lighter now that benchmarks show it doesn't matter, perhaps this should be closed instead.

**Comment by Frooxius** [comment]
No, this absolutely should not be a cloud variable.

We already have a system for settings, we don't need to be introducing new mechanisms for this - it's not a problem that would be blocking this issue.

The issue here isn't with how to store a setting, it's how to enforce it and synchronize across all particle systems in all the worlds.

E.g. say you have a limit of 1000000 particles. Then you ha e a system that has same limit.

This can very quickly exhaust the entire limit, breaking all other particle systems that emit only a few.

There are also issues how to enforce the limit across many particle systems that are all updating independently of each other and on other threads.

That's non-trivial.

Very likely we'd have to compute total limits for each instance with smaller limits being favored and larger being clamped, so each system still limits itself for the exact particle count.

**Comment by epicEaston197** [comment]
When I was thinking of this I imagined the particle limit being apply to each particle System independently Basically just adjusting the "Max particles" field Locally to me whatever the user inputted into the settings menu

>Very likely we'd have to compute total limits for each instance with smaller limits being favored and larger being clamped, so each system still limits itself for the exact particle count.

This solution does make sense though. And has much more nuance and could potentially preserve the intended effect while Keeping the particle count generally lower I also like this solution because when I say "I want a Max of 100,000 particles" I would expect there to be only 100,000 particles Not for that to only apply to each individual system

So yeah this issue does have much more complexities than I originally expected
Looking back at my post This is basically what I wanted

Also the reason why Nuki wanted it to be a cloud variable is because we don't have any easy ways of grabbing a user settings for preferences in cloud variables allowed that so I can understand the need and want for that I do agree with your perspective though 
We already have settings system for this

>...then again, is this still an issue after photon dust? If the particle system is so much lighter now that benchmarks show it doesn't matter, perhaps this should be closed instead.

For context this issue was made right after my graphics card gave drivers out after someone emitted who knows how many particles the same can definitely happen with photon dust Like in Redd's firework world I was acting really stupid because I wanted to see how many fireworks I could emitt but my graphics card did not have any issues I just had a lower frame rate possibly because of the CPU usage or something

**Comment by Redd56** [comment]
Also theres the fact that particle performance can vary wildly. 
You could have over 100k opaque particles that dont have any movement and both your cpu and gpu get barely touched. But throw it to alpha or add in some movement (or worse add collisions to the movement) and your gpu/cpu respectively gets hammered.

**Comment by Frooxius** [comment]
Doing it independently would be much easier to implement, but also doesn't prevent "Spawn 100000 particle systems" kind of thing. Though that can be kinda separate problem.

I think the limit would also need to be defined in terms of geometry. E.g. it's different to have 10000 particles that are just quad, vs 10000 particles that have 10000 triangle mesh particle each.

**Comment by brecert** [comment]
Particle systems could have `MinParticleCountMinimum` and a `MaxParticleCountMaximum` field (with better, less confusing names of course).

In the settings users could adjust a ratio (0%-100%) that would lerp between between those two values universally.

This lets creators still control the intent behind their particles counts while allowing for users to decide what amount of particles to see, which is an important aspect to me.

**Comment by Frooxius** [comment]
I don't think that would solve the problem with people going too crazy with particles, because it requires creators to manually set those limits up so it respects the setting.

**Comment by NukiRaccoon** [comment]
But it WOULD at least give us an option to better optimize what we update or create moving forward, which was what I was thinking when I suggested there be a CloudVar, which is still something people in the community could standardize.

**Comment by Frooxius** [comment]
@NukiRaccoon If it serves a different goal from this issue, it should probably be its own feature request.

**Comment by epicEaston197** [comment]
> [@NukiRaccoon](https://github.com/NukiRaccoon) If it serves a different goal from this issue, it should probably be its own feature request.

I think what Nuki is asking for would be better achieved by a General way of being able to grab a user's preferences Example if someone has Ear muff mode on you'll be able to access that via a component or something to enable a visual

But yeah that's off topic and Nuki should Make it its own issue
